File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        
        q = deque()
        low =0
        high =0
        ans = []
        
        while high < len(nums):
            
            while q and q[-1] < nums[high]:
                q.pop()
                
            q.append(nums[high])
            
            if high - low +1 == k:
                
                ans.append(q[0])
                
                if q[0] == nums[low]:
                    q.popleft()
                low +=1
                
                
            high+=1
            
        return ans
                
                
        

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        
        deque<int>q;
        
        int low =0;
        int high = 0;
        vector<int>ans;
            
        while(high < nums.size()){
            
            while(q.size()>0 and q.back() < nums[high]) q.pop_back();
            q.push_back(nums[high]);
            
            if(high - low +1 == k){
                ans.push_back(q.front());
                if(nums[low] == q.front()) q.pop_front();
                low++;
            }
            
            high++;
        }
        
        return ans;
        
    }
};